File: wear/src/main/java/hibernate/v2/testyourandroid/ui/fragment/TestSensorFragment.java
package hibernate.v2.testyourandroid.ui.fragment;

import android.annotation.SuppressLint;
import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import hibernate.v2.testyourandroid.R;
import hibernate.v2.testyourandroid.model.InfoHeader;
import hibernate.v2.testyourandroid.model.InfoItem;
import hibernate.v2.testyourandroid.ui.adaptor.InfoItemAdaptor;
import hibernate.v2.testyourandroid.helper.SensorHelper;

/**
 * Created by himphen on 21/5/16.
 */
public class TestSensorFragment extends BaseFragment {

	private SensorManager mgr;
	private Sensor sensor;
	private SensorEventListener sensorEventListener = null;

	public static final String ARG_SENSOR_TYPE = "sensorType";

	private int sensorType = 0;

	private String reading = "";
	private InfoItemAdaptor adapter;
	private List<InfoItem> list = new ArrayList<>();

	@BindView(R.id.rvlist)
	RecyclerView recyclerView;

	public static TestSensorFragment newInstance(int sensorType) {
		TestSensorFragment fragment = new TestSensorFragment();
		Bundle args = new Bundle();
		args.putInt(ARG_SENSOR_TYPE, sensorType);
		fragment.setArguments(args);
		return fragment;
	}

	public TestSensorFragment() {
		// Required empty public constructor
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		if (getArguments() != null) {
			sensorType = getArguments().getInt(ARG_SENSOR_TYPE, 0);
		}
	}

	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
	                         Bundle savedInstanceState) {
		// Inflate the layout for this fragment
		View rootView = inflater.inflate(R.layout.fragment_info_listview, container, false);
		ButterKnife.bind(this, rootView);
		return rootView;
	}

	@Override
	public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		recyclerView.setLayoutManager(
				new LinearLayoutManager(mContext,
						RecyclerView.VERTICAL, false)
		);
		init();
	}

	@Override
	public void onPause() {
		super.onPause();
		if (sensor != null && sensorEventListener != null) {
			mgr.unregisterListener(sensorEventListener);
		}
	}

	@Override
	public void onResume() {
		super.onResume();
		if (sensor != null && sensorEventListener != null) {
			mgr.registerListener(sensorEventListener, sensor,
					SensorManager.SENSOR_DELAY_NORMAL);
		}
	}

	private void init() {
		list = new ArrayList<>();
		String[] stringArray = getResources().getStringArray(R.array.test_sensor_string_array);

		mgr = (SensorManager) mContext.getSystemService(Context.SENSOR_SERVICE);

		try {
			sensor = mgr.getDefaultSensor(sensorType);
			if (sensor == null) {
				Toast.makeText(mContext, R.string.dialog_feature_na_message, Toast.LENGTH_LONG).show();
				mContext.finish();
				return;
			}
		} catch (Exception e) {
			Toast.makeText(mContext, R.string.dialog_feature_na_message, Toast.LENGTH_LONG).show();
			mContext.finish();
			return;
		}

		switch (sensorType) {
			case Sensor.TYPE_ACCELEROMETER:
				sensorEventListener = accelerometerListener;
				break;
			case Sensor.TYPE_GRAVITY:
				sensorEventListener = accelerometerListener;
				break;
			case Sensor.TYPE_LIGHT:
				sensorEventListener = lightListener;
				break;
			case Sensor.TYPE_PRESSURE:
				sensorEventListener = pressureListener;
				break;
			case Sensor.TYPE_PROXIMITY:
				sensorEventListener = proximityListener;
				break;
			default:
		}
		InfoItem infoItem;
		for (int i = 0; i < stringArray.length; i++) {
			try {
				switch (sensorType) {
					case Sensor.TYPE_ACCELEROMETER:
						infoItem = new InfoItem(stringArray[i], SensorHelper.getAccelerometerSensorData(i, stringArray.length, reading, sensor));
						break;
					case Sensor.TYPE_GRAVITY:
						infoItem = new InfoItem(stringArray[i], SensorHelper.getGravitySensorData(i, stringArray.length, reading, sensor));
						break;
					case Sensor.TYPE_LIGHT:
						infoItem = new InfoItem(stringArray[i], SensorHelper.getLightSensorData(i, stringArray.length, reading, sensor));
						break;
					case Sensor.TYPE_PRESSURE:
						infoItem = new InfoItem(stringArray[i], SensorHelper.getPressureSensorData(i, stringArray.length, reading, sensor));
						break;
					case Sensor.TYPE_PROXIMITY:
						infoItem = new InfoItem(stringArray[i], SensorHelper.getProximitySensorData(i, stringArray.length, reading, sensor));
						break;
					default:
						infoItem = new InfoItem(stringArray[i], getString(R.string.ui_not_support));
				}
			} catch (Exception e) {
				e.printStackTrace();
				infoItem = new InfoItem(stringArray[i], getString(R.string.ui_not_support));
			}

			list.add(infoItem);
		}

		adapter = new InfoItemAdaptor(list);
		adapter.setHeader(new InfoHeader(mContext.getTitle().toString()));
		recyclerView.setAdapter(adapter);
	}

	private SensorEventListener accelerometerListener = new SensorEventListener() {
		@Override
		public void onAccuracyChanged(Sensor arg0, int arg1) {
		}

		@SuppressLint("DefaultLocale")
		@Override
		public void onSensorChanged(SensorEvent event) {
			reading = "X: " + String.format("%1.4f", event.values[0]) + " m/s²\nY: "
					+ String.format("%1.4f", event.values[1]) + " m/s²\nZ: "
					+ String.format("%1.4f", event.values[2]) + " m/s²";
			list.get(0).setContentText(reading);
			adapter.notifyDataSetChanged();
		}
	};

	private SensorEventListener lightListener = new SensorEventListener() {
		@Override
		public void onAccuracyChanged(Sensor arg0, int arg1) {
		}

		@Override
		public void onSensorChanged(SensorEvent event) {
			reading = String.valueOf(event.values[0]) + " lux";
			list.get(0).setContentText(reading);
			adapter.notifyDataSetChanged();
		}
	};

	private SensorEventListener pressureListener = new SensorEventListener() {
		@Override
		public void onAccuracyChanged(Sensor arg0, int arg1) {
		}

		@Override
		public void onSensorChanged(SensorEvent event) {
			reading = String.valueOf(event.values[0]) + " hPa";
			list.get(0).setContentText(reading);
			adapter.notifyDataSetChanged();
		}
	};

	private SensorEventListener proximityListener = new SensorEventListener() {
		@Override
		public void onAccuracyChanged(Sensor arg0, int arg1) {
		}

		@SuppressLint("DefaultLocale")
		@Override
		public void onSensorChanged(SensorEvent event) {
			reading = String.format("%1.2f", event.values[0]) + " cm";
			list.get(0).setContentText(reading);
			adapter.notifyDataSetChanged();
		}
	};

}

File: wear/src/main/java/hibernate/v2/testyourandroid/ui/test/TestSensorFragment.kt
package hibernate.v2.testyourandroid.ui.test

import android.annotation.SuppressLint
import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import hibernate.v2.testyourandroid.R
import hibernate.v2.testyourandroid.helper.SensorHelper
import hibernate.v2.testyourandroid.model.InfoHeader
import hibernate.v2.testyourandroid.model.InfoItem
import hibernate.v2.testyourandroid.ui.base.InfoItemAdapter
import hibernate.v2.testyourandroid.ui.base.BaseFragment
import kotlinx.android.synthetic.main.fragment_info_listview.*
import java.util.ArrayList

/**
 * Created by himphen on 21/5/16.
 */
class TestSensorFragment : BaseFragment() {
    private lateinit var sensorManager: SensorManager
    private var sensor: Sensor? = null
    private var sensorEventListener: SensorEventListener? = null
    private var sensorType = 0
    private var reading = ""
    private lateinit var adapter: InfoItemAdapter
    private var list: MutableList<InfoItem> = ArrayList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        sensorType = arguments?.getInt(ARG_SENSOR_TYPE, 0) ?: 0
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? { // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_info_listview, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        rvlist.layoutManager = LinearLayoutManager(context)
        init()
    }

    override fun onPause() {
        super.onPause()
        if (sensor != null && sensorEventListener != null) {
            sensorManager.unregisterListener(sensorEventListener)
        }
    }

    override fun onResume() {
        super.onResume()
        if (sensor != null && sensorEventListener != null) {
            sensorManager.registerListener(sensorEventListener, sensor,
                    SensorManager.SENSOR_DELAY_NORMAL)
        }
    }

    private fun init() {
        list = ArrayList()
        val stringArray = resources.getStringArray(R.array.test_sensor_string_array)
        sensorManager = context!!.getSystemService(Context.SENSOR_SERVICE) as SensorManager
        try {
            sensor = sensorManager.getDefaultSensor(sensorType)
            if (sensor == null) {
                Toast.makeText(context, R.string.dialog_feature_na_message, Toast.LENGTH_LONG).show()
                activity?.finish()
                return
            }
        } catch (e: Exception) {
            Toast.makeText(context, R.string.dialog_feature_na_message, Toast.LENGTH_LONG).show()
            activity?.finish()
            return
        }
        when (sensorType) {
            Sensor.TYPE_ACCELEROMETER, Sensor.TYPE_GRAVITY -> sensorEventListener = accelerometerListener
            Sensor.TYPE_LIGHT -> sensorEventListener = lightListener
            Sensor.TYPE_PRESSURE -> sensorEventListener = pressureListener
            Sensor.TYPE_PROXIMITY -> sensorEventListener = proximityListener
            else -> {
            }
        }
        var infoItem: InfoItem
        for (i in stringArray.indices) {
            infoItem = try {
                when (sensorType) {
                    Sensor.TYPE_ACCELEROMETER -> InfoItem(stringArray[i], SensorHelper.getAccelerometerSensorData(i, stringArray.size, reading, sensor!!))
                    Sensor.TYPE_GRAVITY -> InfoItem(stringArray[i], SensorHelper.getGravitySensorData(i, stringArray.size, reading, sensor!!))
                    Sensor.TYPE_LIGHT -> InfoItem(stringArray[i], SensorHelper.getLightSensorData(i, stringArray.size, reading, sensor!!))
                    Sensor.TYPE_PRESSURE -> InfoItem(stringArray[i], SensorHelper.getPressureSensorData(i, stringArray.size, reading, sensor!!))
                    Sensor.TYPE_PROXIMITY -> InfoItem(stringArray[i], SensorHelper.getProximitySensorData(i, stringArray.size, reading, sensor!!))
                    else -> InfoItem(stringArray[i], getString(R.string.ui_not_support))
                }
            } catch (e: Exception) {
                e.printStackTrace()
                InfoItem(stringArray[i], getString(R.string.ui_not_support))
            }
            list.add(infoItem)
        }
        adapter = InfoItemAdapter(list)
        adapter.header = InfoHeader(activity?.title.toString())
        rvlist.adapter = adapter
    }

    private val accelerometerListener: SensorEventListener = object : SensorEventListener {
        override fun onAccuracyChanged(arg0: Sensor, arg1: Int) {}
        @SuppressLint("DefaultLocale")
        override fun onSensorChanged(event: SensorEvent) {
            reading = ("X: " + String.format("%1.4f", event.values[0]) + " m/s²\nY: "
                    + String.format("%1.4f", event.values[1]) + " m/s²\nZ: "
                    + String.format("%1.4f", event.values[2]) + " m/s²")
            list[0].contentText = reading
            adapter.notifyDataSetChanged()
        }
    }
    private val lightListener: SensorEventListener = object : SensorEventListener {
        override fun onAccuracyChanged(arg0: Sensor, arg1: Int) {}
        override fun onSensorChanged(event: SensorEvent) {
            reading = event.values[0].toString() + " lux"
            list[0].contentText = reading
            adapter.notifyDataSetChanged()
        }
    }
    private val pressureListener: SensorEventListener = object : SensorEventListener {
        override fun onAccuracyChanged(arg0: Sensor, arg1: Int) {}
        override fun onSensorChanged(event: SensorEvent) {
            reading = event.values[0].toString() + " hPa"
            list[0].contentText = reading
            adapter.notifyDataSetChanged()
        }
    }
    private val proximityListener: SensorEventListener = object : SensorEventListener {
        override fun onAccuracyChanged(arg0: Sensor, arg1: Int) {}
        @SuppressLint("DefaultLocale")
        override fun onSensorChanged(event: SensorEvent) {
            reading = String.format("%1.2f", event.values[0]) + " cm"
            list[0].contentText = reading
            adapter.notifyDataSetChanged()
        }
    }

    companion object {
        const val ARG_SENSOR_TYPE = "sensorType"
        fun newInstance(sensorType: Int): TestSensorFragment {
            val fragment = TestSensorFragment()
            val args = Bundle()
            args.putInt(ARG_SENSOR_TYPE, sensorType)
            fragment.arguments = args
            return fragment
        }
    }
}